### PR TITLE
[Cart] Cart with shipping fee and tax scenarios

### PR DESCRIPTION
**Pull request by Zales0123**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Fixed tickets |
| License       | MIT
| Doc PR        |

Based on https://github.com/Sylius/Sylius/pull/4002 - added scenarios for taxed shipping methods.

### CODE REVIEW

**Comment by michalmarcinkowski** [comment]
@Zales0123 needs CS fixes, rebase and squash :)

File: src/Sylius/Behat/Context/Setup/TaxContext.php [hunk 1]
```diff
@@ -104,6 +102,26 @@ public function storeHasTaxRateWithinZone($taxRateName, $taxRateAmount, $taxCate
...
+        try {
+            return $this->getTaxCategoryByName($taxCategoryName);
+        } catch (\InvalidArgumentException $exception) {
+            $taxCategory = $this->taxCategoryFactory->createNew();
+            $taxCategory->setName($taxCategoryName);
+            $taxCategory->setCode($this->getCodeFromName($taxCategoryName));
+
+            $this->taxCategoryRepository->add($taxCategory);
+
+            return $taxCategory;
```
Comment (michalmarcinkowski) [inline src/Sylius/Behat/Context/Setup/TaxContext.php]: Extract all these to `createTaxCategory` method.

File: src/Sylius/Behat/Context/Setup/TaxContext.php [hunk 2]
```diff
@@ -104,6 +102,26 @@ public function storeHasTaxRateWithinZone($taxRateName, $taxRateAmount, $taxCate
     }
 
     /**
+     * @param string $taxCategoryName
+     *
+     * @return TaxCategoryInterface
+     */
+    private function getTaxCategory($taxCategoryName)
```
Comment (michalmarcinkowski) [inline src/Sylius/Behat/Context/Setup/TaxContext.php]: `getOrCreateTaxCategory`

File: src/Sylius/Component/Core/Test/Services/DefaultFranceChannelFactory.php [hunk 3]
```diff
@@ -94,15 +99,20 @@ public function __construct(
...
         $defaultData['channel'] = $this->createChannel();
-        $defaultData['country'] = $this->createCountry();
         $defaultData['zone_member'] = $this->createZoneMember();
         $defaultData['zone'] = $this->createZone($defaultData['zone_member']);
 
         $this->channelRepository->add($defaultData['channel']);
-        $this->countryRepository->add($defaultData['country']);
+
+        foreach ($this->defaultCountries as $country) {
+            $this->countryRepository->add($this->createCountry($country));
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Core/Test/Services/DefaultFranceChannelFactory.php]: It's not a proper place to create default countries.

File: features/shop/apply_correct_shipping_fee_with_taxes_on_order.feature [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+        And default currency is "EUR"
+        And default tax zone is "EU"
+        And there is user "john@example.com" identified by "password123"
+        And store has "EU VAT" tax rate of 23% for "Clothes" within "EU" zone
+        And store has "Low tax" tax rate of 10% for "Clothes" for the rest of the world
+        And store has a product "PHP T-Shirt" priced at "€100.00"
+        And store has "DHL" shipping method with "€10.00" fee within "EU" zone
+        And store has "DHL-World" shipping method with "€20.00" fee for the rest of the world
+        And shipping method "DHL" belongs to "Clothes" tax category
+        And shipping method "DHL-World" belongs to "Clothes" tax category
```
Comment (michalmarcinkowski) [inline features/shop/apply_correct_shipping_fee_with_taxes_on_order.feature]: `belongs to "Clothes" tax category` Clothes? :smile:

File: features/shop/apply_correct_shipping_fee_with_taxes_on_order.feature [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+        And store allows paying offline
+        And I am logged in as "john@example.com"
+
+    Scenario: Proper shipping fee and tax
+        Given I am logged in as "john@example.com"
+        And I add product "PHP T-Shirt" to the cart
```
Comment (michalmarcinkowski) [inline features/shop/apply_correct_shipping_fee_with_taxes_on_order.feature]: `I have product ...`

File: features/shop/apply_correct_shipping_fee_with_taxes_on_order.feature [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+        When I proceed selecting "DHL" shipping method
+        Then my cart shipping fee should be "€12.30"
+        And my cart taxes should be "€2.30"
+        Then my cart total should be "€112.30"
+
+    Scenario: Proper shipping fee and tax after addressing
+        Given I am logged in as "john@example.com"
+        And I add product "PHP T-Shirt" to the cart
+        And I proceed selecting "DHL" shipping method and "Australia" as shipping country
```
Comment (michalmarcinkowski) [inline features/shop/apply_correct_shipping_fee_with_taxes_on_order.feature]: "DHL" shipping method is not available for "Australia", we should select "DHL-World". 

**When** I proceed ...

File: src/Sylius/Behat/Context/Setup/ShippingContext.php [hunk 7]
```diff
@@ -69,18 +71,30 @@ public function getShippingMethodByName($shippingMethodName)
...
+        $this->createShippingMethod($shippingMethodName, $zone, 'en', ['amount' => $this->getFeeFromString($fee)]);
+    }
+
+    /**
+     * @Given /^(shipping method "[^"]+") belongs to ("[^"]+" tax category)$/
+     */
+    public function productBelongsToTaxCategory(ShippingMethodInterface $shippingMethod, TaxCategoryInterface $taxCategory)
+    {
+        $shippingMethod->setTaxCategory($taxCategory);
+        $this->shippingMethodRepository->add($shippingMethod);
```
Comment (michalmarcinkowski) [inline src/Sylius/Behat/Context/Setup/ShippingContext.php]: `$manager->flush`?

File: src/Sylius/Behat/Context/Setup/ZoneContext.php [hunk 8]
```diff
@@ -73,6 +74,19 @@ public function getZoneByCode($zone)
     }
 
     /**
+     * @Transform /^rest of the world$/
+     * @Transform /^the rest of the world$/
+     */
+    public function castRestOfTheWorldToZone()
+    {
+        if (null === $zone = $this->zoneRepository->findOneBy(['code' => 'RoW'])) {
```
Comment (michalmarcinkowski) [inline src/Sylius/Behat/Context/Setup/ZoneContext.php]: Extract variable before `if`. Rename method to `get...`

File: src/Sylius/Behat/Context/Ui/CheckoutContext.php [hunk 9]
```diff
@@ -127,6 +128,9 @@ public function iProceedSelectingShippingMethod($shippingMethod)
 
         $checkoutShippingPage = $this->getPage('Checkout\CheckoutShippingStep');
         $checkoutShippingPage->selectShippingMethod($shippingMethod);
+
+        $checkoutPaymentPage = $this->getPage('Checkout\CheckoutPaymentStep');
+        $checkoutPaymentPage->selectPaymentMethod('Offline');
```
Comment (michalmarcinkowski) [inline src/Sylius/Behat/Context/Ui/CheckoutContext.php]: Why we need to select payment at this step?

File: features/shop/apply_correct_shipping_fee_on_order.feature [hunk 10]
```diff
@@ -10,6 +10,7 @@ Feature: Apply correct shipping fee on order
         And store has a product "PHP T-Shirt" priced at "€100.00"
         And store has "DHL" shipping method with "€10.00" fee
         And store has "FedEx" shipping method with "€30.00" fee
+        And store allows paying offline
```
Comment (michalmarcinkowski) [inline features/shop/apply_correct_shipping_fee_on_order.feature]: Is it needed?

File: features/shop/apply_correct_shipping_fee_with_taxes_on_order.feature [hunk 11]
```diff
@@ -0,0 +1,38 @@
+@ui-cart
+Feature: Apply correct shipping fee with taxes on order
+    In order to pay proper amount when buying goods and choosing taxes shipping method
```
Comment (michalmarcinkowski) [inline features/shop/apply_correct_shipping_fee_with_taxes_on_order.feature]: `In order to pay proper amount for shipping`

File: features/shop/apply_correct_shipping_fee_with_taxes_on_order.feature [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+        And store allows paying offline
+        And I am logged in as "john@example.com"
+
+    Scenario: Proper shipping fee and tax
+        Given I am logged in as "john@example.com"
+        And I add product "PHP T-Shirt" to the cart
+        When I proceed selecting "DHL" shipping method
+        Then my cart shipping fee should be "€12.30"
+        And my cart taxes should be "€2.30"
+        Then my cart total should be "€112.30"
```
Comment (michalmarcinkowski) [inline features/shop/apply_correct_shipping_fee_with_taxes_on_order.feature]: Everywhere else we have the following order:
```
Then cart total 
And cart taxes
And cart shipping
```

File: features/shop/apply_correct_shipping_fee_with_taxes_on_order.feature (hunk 4 above)
Comment (pjedrzejewski) [inline features/shop/apply_correct_shipping_fee_with_taxes_on_order.feature]: Shipping Services

File: features/shop/apply_correct_shipping_fee_on_order.feature (hunk 10 above)
Comment (Zales0123) [inline features/shop/apply_correct_shipping_fee_on_order.feature]: Unfortunately yes. Step ``When I proceed selecting "..." shipping method`` requires default "Offline" payment method, as taxes (for example for shipping) are applied after "payment" checkout step (``sylius.checkout.payment.pre_complete``);
In the future, when we add proper callbacks on order state it will be unnecessary, but know it would force us to deal with listeners priorities ;/

File: src/Sylius/Behat/Context/Ui/CheckoutContext.php (hunk 9 above)
Comment (Zales0123) [inline src/Sylius/Behat/Context/Ui/CheckoutContext.php]: See (this comment)[https://github.com/Sylius/Sylius/pull/4010#discussion_r51862971]

File: features/shop/apply_correct_shipping_fee_on_order.feature [hunk 13]
```diff
@@ -5,11 +5,13 @@ Feature: Apply correct shipping fee on order
     I want to have shipping fee applied based on chosen shipping method
 
     Background:
-        Given the store is operating on a single "France" channel
+        Given the store is operating on a single channel
+        And store ships to "Australia", "France" and "United States"
```
Comment (pamil) [inline features/shop/apply_correct_shipping_fee_on_order.feature]: IMHO we should always use **the** store

File: features/shop/apply_correct_shipping_fee_with_taxes_on_order.feature [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+
+    Scenario: Proper shipping fee and tax after addressing
+        Given I am logged in as "john@example.com"
+        And I add product "PHP T-Shirt" to the cart
+        And I proceed selecting "DHL" shipping method and "Australia" as shipping country
+        Then my cart shipping fee should be "€22.00"
+        And my cart taxes should be "€2.00"
+        Then my cart total should be "€122.00"
```
Comment (pamil) [inline features/shop/apply_correct_shipping_fee_with_taxes_on_order.feature]: And

File: features/shop/apply_correct_shipping_fee_on_order.feature (hunk 13 above)
Comment (Zales0123) [inline features/shop/apply_correct_shipping_fee_on_order.feature]: Aggreed :+1: however, there are many steps where it should be changed, so I think we should do this in separate PR, wdyt?

Comment (pamil) [inline features/shop/apply_correct_shipping_fee_on_order.feature]: :+1:

File: src/Sylius/Component/Core/Test/Services/DefaultCountriesFactoryInterface.php [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Component\Core\Test\Services;
+
+/**
+ * @author Mateusz Zalewski <mateusz.zalewski@lakion.com>
+ */
+interface DefaultCountriesFactoryInterface
```
Comment (pamil) [inline src/Sylius/Component/Core/Test/Services/DefaultCountriesFactoryInterface.php]: It looks more like a provider, it also adds them to repository. The "default" keyword is not needed here IMO.

Also, why don't we create `CountryProviderInterface::provide($countryCode)` that:
  - returns Country object if it exists in the repository
  - creates a new Country and adds it into repository

This way we get rid of that array parameter that doesn't look so good and we won't miss the additional features that comes with this "factory". It will also allow us to do something more with that Country object.

File: src/Sylius/Behat/Context/Ui/CheckoutContext.php [hunk 16]
```diff
@@ -108,15 +108,16 @@ public function iProceedSelectingOfflinePaymentMethod($paymentMethodName)
...
      */
-    public function iProceedSelectingShippingMethod($shippingMethod)
+    public function iProceedSelectingShippingMethodAndShippingCountry($shippingMethod, $shippingCountry = null)
     {
         $checkoutAddressingPage = $this->getPage('Checkout\CheckoutAddressingStep')->open();
         $addressingDetails = [
             'firstName' => 'John',
             'lastName' => 'Doe',
-            'country' => 'France',
+            'country' => (null === $shippingCountry) ? 'France' : $shippingCountry,
```
Comment (pamil) [inline src/Sylius/Behat/Context/Ui/CheckoutContext.php]: `$shippingCountry ?: 'France'`

File: src/Sylius/Behat/Context/Setup/TaxContext.php (hunk 1 above)
Comment (pamil) [inline src/Sylius/Behat/Context/Setup/TaxContext.php]: Or even some `TaxCategoryProvider` class.

File: src/Sylius/Behat/Context/Setup/ShippingContext.php [hunk 17]
```diff
@@ -69,18 +71,30 @@ public function getShippingMethodByName($shippingMethodName)
...
+    public function storeHasShippingMethodWithFee($shippingMethodName, $fee, ZoneInterface $zone = null)
     {
-        $this->createShippingMethod($shippingMethodName, 'en', ['amount' => $this->getFeeFromString($fee)]);
+        $this->createShippingMethod($shippingMethodName, $zone, 'en', ['amount' => $this->getFeeFromString($fee)]);
+    }
+
+    /**
+     * @Given /^(shipping method "[^"]+") belongs to ("[^"]+" tax category)$/
+     */
+    public function productBelongsToTaxCategory(ShippingMethodInterface $shippingMethod, TaxCategoryInterface $taxCategory)
```
Comment (pamil) [inline src/Sylius/Behat/Context/Setup/ShippingContext.php]: shippingMethodBelongsToTaxCategory?

File: src/Sylius/Behat/Context/Setup/ShippingContext.php [hunk 18]
```diff
@@ -69,18 +71,30 @@ public function getShippingMethodByName($shippingMethodName)
...
+    public function storeShipsEverythingForFree(ZoneInterface $zone = null)
     {
-        $this->createShippingMethod('Free');
+        $this->createShippingMethod('Free', $zone);
     }
 
     /**
      * @Given /^store has "([^"]*)" shipping method with "(?:€|£|\$)([^"]*)" fee$/
+     * @Given /^store has "([^"]*)" shipping method with "(?:€|£|\$)([^"]*)" fee within ("([^"]*)" zone)$/
+     * @Given /^store has "([^"]*)" shipping method with "(?:€|£|\$)([^"]*)" fee for (the rest of the world)$/
```
Comment (pamil) [inline src/Sylius/Behat/Context/Setup/ShippingContext.php]: `"(?:€|£|\$)([^"]*)"` is widely used, can we make something to reuse it?

File: src/Sylius/Behat/Context/Setup/ShippingContext.php [hunk 19]
```diff
@@ -69,18 +71,30 @@ public function getShippingMethodByName($shippingMethodName)
 
     /**
      * @Given store ships everything for free
+     * @Given /^store ships everything for free within ("([^"]*)" zone)$/
```
Comment (pamil) [inline src/Sylius/Behat/Context/Setup/ShippingContext.php]: Double matching `()` may create some issues in the future.

Comment (pamil) [inline src/Sylius/Behat/Context/Setup/ShippingContext.php]: Anyway, for those simple steps I prefer not using regexps:
`@Given store ships everything for free within :zone zone`
And then zone transformer with `@Transform :zone`

File: src/Sylius/Behat/Context/Ui/CheckoutContext.php (hunk 16 above)
Comment (pjedrzejewski) [inline src/Sylius/Behat/Context/Ui/CheckoutContext.php]: :+1:

File: src/Sylius/Behat/Context/Setup/ShippingContext.php (hunk 18 above)
Comment (Zales0123) [inline src/Sylius/Behat/Context/Setup/ShippingContext.php]: Roger that, but for now I cannot came up with any better idea ;/

File: src/Sylius/Component/Core/Test/Services/DefaultCountriesFactoryInterface.php (hunk 15 above)
Comment (Zales0123) [inline src/Sylius/Component/Core/Test/Services/DefaultCountriesFactoryInterface.php]: I think question is - do we really need a separate service **for now**? The only feature we're trying to achieve is configuring store shipping-available countries. There is no need for provider described by you, but also I suppose there is no need for factory I've implemented :smile:
So my idea is to get rid of any additional service and configure shipping countries for store in ``ChannelContext`` (or event separate ``CountryContext``?). @pamil WDYT?

File: src/Sylius/Behat/Context/Setup/ShippingContext.php (hunk 19 above)
Comment (michalmarcinkowski) [inline src/Sylius/Behat/Context/Setup/ShippingContext.php]: :+1:

File: features/shop/apply_correct_shipping_fee_on_order.feature (hunk 10 above)
Comment (michalmarcinkowski) [inline features/shop/apply_correct_shipping_fee_on_order.feature]: We should change it ASAP :)

File: features/shop/apply_correct_shipping_fee_with_taxes_on_order.feature [hunk 20]
```diff
@@ -0,0 +1,41 @@
...
+        And there is "EU" zone containing all members of European Union
+        And there is rest of the world zone containing all other countries
+        And store ships to "France" and "Australia"
+        And default currency is "EUR"
+        And default tax zone is "EU"
+        And there is user "john@example.com" identified by "password123"
+        And store has "EU VAT" tax rate of 23% for "Clothes" within "EU" zone
+        And store has "Low tax" tax rate of 10% for "Clothes" for the rest of the world
+        And store has "Shipping EU VAT" tax rate of 23% for "Shipping Services" within "EU" zone
+        And store has "Shipping Low tax" tax rate of 10% for "Shipping Services World" for the rest of the world
```
Comment (michalmarcinkowski) [inline features/shop/apply_correct_shipping_fee_with_taxes_on_order.feature]: The tax category can also be "Shipping Services" here. Thanks to zone matching the proper tax will be applied for Australia anyway.

File: features/shop/apply_correct_shipping_fee_with_taxes_on_order.feature [hunk 21]
```diff
@@ -0,0 +1,41 @@
...
+        And store has a product "PHP T-Shirt" priced at "€100.00"
+        And store has "DHL" shipping method with "€10.00" fee within "EU" zone
+        And store has "DHL-World" shipping method with "€20.00" fee for the rest of the world
+        And shipping method "DHL" belongs to "Shipping Services" tax category
+        And shipping method "DHL-World" belongs to "Shipping Services World" tax category
+        And store allows paying offline
+        And I am logged in as "john@example.com"
+
+    Scenario: Proper shipping fee and tax
+        Given I am logged in as "john@example.com"
```
Comment (michalmarcinkowski) [inline features/shop/apply_correct_shipping_fee_with_taxes_on_order.feature]: Redundant, it's the last step of the background

File: features/shop/apply_correct_shipping_fee_with_taxes_on_order.feature [hunk 22]
```diff
@@ -0,0 +1,41 @@
...
+        And I have product "PHP T-Shirt" in the cart
+        When I proceed selecting "DHL" shipping method
+        Then my cart total should be "€112.30"
+        And my cart taxes should be "€2.30"
+        And my cart shipping fee should be "€12.30"
+
+    Scenario: Proper shipping fee and tax after addressing
+        Given I am logged in as "john@example.com"
```
Comment (michalmarcinkowski) [inline features/shop/apply_correct_shipping_fee_with_taxes_on_order.feature]: Same here.

File: src/Sylius/Component/Core/Test/Services/DefaultChannelFactory.php [hunk 23]
```diff
@@ -22,17 +21,16 @@
 /**
  * @author Arkadiusz Krakowiak <arkadiusz.krakowiak@lakion.com>
  */
-class DefaultFranceChannelFactory implements DefaultStoreDataInterface
+class DefaultChannelFactory implements DefaultStoreDataInterface
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Core/Test/Services/DefaultChannelFactory.php]: I would still call it `DefaultFranceChannelFactory` since it creates french channel :)

File: src/Sylius/Component/Core/Test/Services/DefaultChannelFactory.php [hunk 24]
```diff
@@ -66,43 +59,38 @@ class DefaultFranceChannelFactory implements DefaultStoreDataInterface
...
         $this->zoneFactory = $zoneFactory;
     }
 
+    /**
+     * {@inheritdoc}
+     */
     public function create()
     {
         $defaultData['channel'] = $this->createChannel();
-        $defaultData['country'] = $this->createCountry();
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Core/Test/Services/DefaultChannelFactory.php]: I would leave this.

File: src/Sylius/Behat/Context/Setup/ChannelContext.php [hunk 25]
```diff
@@ -21,33 +23,82 @@
...
     ) {
-        $this->defaultFranceChannelFactory = $defaultFranceChannelFactory;
         $this->sharedStorage = $sharedStorage;
+        $this->defaultChannelFactory = $defaultChannelFactory;
+        $this->defaultCountriesFactory = $defaultCountriesFactory;
     }
 
     /**
-     * @Given the store is operating on a single "France" channel
+     * @Given the store is operating on a single channel
```
Comment (michalmarcinkowski) [inline src/Sylius/Behat/Context/Setup/ChannelContext.php]: I would map both steps to this function and in both cases create "France" channel

File: src/Sylius/Component/Core/Test/Services/DefaultCountriesFactoryInterface.php (hunk 15 above)
Comment (michalmarcinkowski) [inline src/Sylius/Component/Core/Test/Services/DefaultCountriesFactoryInterface.php]: There can be places where we could make use of 'providers', so let's discuss it tomorrow! :)

**Comment by michalmarcinkowski** [comment]
@Zales0123 please rebase and apply all fixes.

File: src/Sylius/Behat/Context/Setup/ChannelContext.php (hunk 25 above)
Comment (Zales0123) [inline src/Sylius/Behat/Context/Setup/ChannelContext.php]: After conversation with @pjedrzejewski we agrees, that there is need of refactoring all country-zone-channel building system, but it should be done in separate PR ;)

Comment (michalmarcinkowski) [inline src/Sylius/Behat/Context/Setup/ChannelContext.php]: :+1:

File: src/Sylius/Behat/spec/Context/Setup/ZoneContextSpec.php [hunk 26]
```diff
@@ -62,8 +63,28 @@ function it_creates_eu_zone_with_european_zone_members(
...
+
+        $this->thereIsRestOfTheWorldZoneContainingAllOtherCountries();
+    }
+
+    function it_sets_default_zone(
+        $zoneRepository,
+        $settingsManager,
+        Settings $settings,
+        ZoneInterface $zone
+    ) {
```
Comment (michalmarcinkowski) [inline src/Sylius/Behat/spec/Context/Setup/ZoneContextSpec.php]: This should fit in one line, doesn't it?

File: features/shop/apply_correct_shipping_fee_with_taxes_on_order.feature [hunk 27]
```diff
@@ -0,0 +1,39 @@
+@ui-cart
+Feature: Apply correct shipping fee with taxes on order
+    In order to pay proper amount for shipping
+    As a Customer
+    I want to have correct shipping fees and taxes applied to my order
+
+    Background:
+        Given the store is operating on a single channel
```
Comment (pjedrzejewski) [inline features/shop/apply_correct_shipping_fee_with_taxes_on_order.feature]: For a separate PR, but we use present simple tense everywhere, so this should be:

```
Given the store operates on a single channel
```

File: features/shop/apply_correct_shipping_fee_with_taxes_on_order.feature [hunk 28]
```diff
@@ -0,0 +1,39 @@
...
+        And there is "EU" zone containing all members of European Union
+        And there is rest of the world zone containing all other countries
+        And store ships to "France" and "Australia"
+        And default currency is "EUR"
+        And default tax zone is "EU"
+        And there is user "john@example.com" identified by "password123"
```
Comment (pjedrzejewski) [inline features/shop/apply_correct_shipping_fee_with_taxes_on_order.feature]: I think this should be before ``I am logged in...`` step.

File: features/shop/apply_correct_shipping_fee_with_taxes_on_order.feature [hunk 29]
```diff
@@ -0,0 +1,39 @@
...
+    Scenario: Proper shipping fee and tax
+        Given I have product "PHP T-Shirt" in the cart
+        When I proceed selecting "DHL" shipping method
+        Then my cart total should be "€112.30"
+        And my cart taxes should be "€2.30"
+        And my cart shipping fee should be "€12.30"
+
+    Scenario: Proper shipping fee and tax after addressing
+        Given I have product "PHP T-Shirt" in the cart
+        When I proceed selecting "DHL-World" shipping method and "Australia" as shipping country
```
Comment (pjedrzejewski) [inline features/shop/apply_correct_shipping_fee_with_taxes_on_order.feature]: I think this should be the other way around, we first decide where to ship and then how.

```
When I proceed selecting "Australia" as shipping country with "DHL-World" method
```

File: src/Sylius/Behat/Context/Setup/ShippingContext.php [hunk 30]
```diff
@@ -69,18 +80,30 @@ public function getShippingMethodByName($shippingMethodName)
...
+
+    /**
+     * @Given /^(shipping method "[^"]+") belongs to ("[^"]+" tax category)$/
      */
-    public function storeHasShippingMethodWithFee($shippingMethodName, $fee)
+    public function shippingMethodBelongsToTaxCategory(ShippingMethodInterface $shippingMethod, TaxCategoryInterface $taxCategory)
     {
-        $this->createShippingMethod($shippingMethodName, 'en', ['amount' => $this->getFeeFromString($fee)]);
+        $shippingMethod->setTaxCategory($taxCategory);
+        $this->shippingMethodManager->flush($shippingMethod);
```
Comment (pjedrzejewski) [inline src/Sylius/Behat/Context/Setup/ShippingContext.php]: We should not flush a specific entity as it usually leads to unexpected results because of missing cascading, etc.

Comment (pjedrzejewski) [inline src/Sylius/Behat/Context/Setup/ShippingContext.php]: Just flush without arguments here.

File: src/Sylius/Behat/Context/Setup/TaxContext.php [hunk 31]
```diff
@@ -71,31 +72,28 @@ public function __construct(
...
+            throw new \InvalidArgumentException('Tax category with name "'.$taxCategoryName.'" does not exist');
         }
 
         return $taxCategory;
     }
 
     /**
-     * @Given /^store has "([^"]+)" tax rate of ([^"]+)% for "([^"]+)" within ("([^"]+)" zone)$/
+     * @Given store has :taxRateName tax rate of :taxRateAmount% for :taxCategoryName within :zone zone
+     * @Given /^store has "([^"]+)" tax rate of ([^"]+)% for "([^"]+)" for (the rest of the world)$/
```
Comment (pjedrzejewski) [inline src/Sylius/Behat/Context/Setup/TaxContext.php]: Any reason to use ``:taxRateName`` and ``"([^"]+)"`` in these 2 step definitions?

File: src/Sylius/Behat/Context/Setup/ZoneContext.php [hunk 32]
```diff
@@ -73,6 +75,20 @@ public function getZoneByCode($zone)
...
 
     /**
+     * @Transform /^rest of the world$/
+     * @Transform /^the rest of the world$/
+     */
+    public function getRestOfTheWorldZone()
+    {
+        $zone = $this->zoneRepository->findOneBy(['code' => 'RoW']);
+        if (null === $zone) {
+            throw new \Exception('Rest of the world zone does not exist');
```
Comment (pjedrzejewski) [inline src/Sylius/Behat/Context/Setup/ZoneContext.php]: Missing dot at the end of exception message.

File: src/Sylius/Behat/Context/Ui/CheckoutContext.php [hunk 33]
```diff
@@ -105,15 +105,16 @@ public function iProceedSelectingOfflinePaymentMethod($paymentMethodName)
...
      */
-    public function iProceedSelectingShippingMethod($shippingMethodName)
+    public function iProceedSelectingShippingMethodAndShippingCountry($shippingMethodName, $shippingCountry = null)
     {
         $this->checkoutAddressingStep->open();
         $this->checkoutAddressingStep->fillAddressingDetails([
             'firstName' => 'John',
             'lastName' => 'Doe',
-            'country' => 'France',
+            'country' => $shippingCountry ?: 'France',
```
Comment (pjedrzejewski) [inline src/Sylius/Behat/Context/Ui/CheckoutContext.php]: Later we should refactor to reflect following logic: If country is not specified, select first available. We should avoid hardcoding France, especially that we removed this "single France channel ..." thingy. :)

File: src/Sylius/Behat/spec/Context/Setup/ChannelContextSpec.php [hunk 34]
```diff
@@ -34,12 +45,12 @@ function it_is_initializable()
         $this->shouldHaveType('Sylius\Behat\Context\Setup\ChannelContext');
     }
 
-    function it_is_context()
+    function it_implements_context()
```
Comment (pjedrzejewski) [inline src/Sylius/Behat/spec/Context/Setup/ChannelContextSpec.php]: I think we should still add ``interface`` here.

File: src/Sylius/Behat/spec/Context/Setup/ChannelContextSpec.php [hunk 35]
```diff
@@ -34,12 +45,12 @@ function it_is_initializable()
...
     {
-        $this->shouldImplement('Behat\Behat\Context\Context');
+        $this->shouldImplement(Context::class);
     }
 
-    function it_sets_default_france_channel_in_to_shared_storage(
+    function it_sets_default_channel_in_to_shared_storage(
```
Comment (pjedrzejewski) [inline src/Sylius/Behat/spec/Context/Setup/ChannelContextSpec.php]: in the shared storage

File: src/Sylius/Behat/spec/Context/Setup/ChannelContextSpec.php [hunk 36]
```diff
@@ -49,6 +60,26 @@ function it_sets_default_france_channel_in_to_shared_storage(
         $defaultFranceChannelFactory->create()->willReturn($defaultData);
         $sharedStorage->setClipboard($defaultData)->shouldBeCalled();
 
-        $this->thatStoreIsOperatingOnASingleFranceChannel();
+        $this->thatStoreIsOperatingOnASingleChannel();
+    }
+
+    function it_configures_shipping_available_countries(
```
Comment (pjedrzejewski) [inline src/Sylius/Behat/spec/Context/Setup/ChannelContextSpec.php]: it configures shipping destination countries

File: src/Sylius/Behat/spec/Context/Setup/ZoneContextSpec.php [hunk 37]
```diff
@@ -79,4 +100,32 @@ function it_throws_exception_if_zone_with_given_code_does_not_exist($zoneReposit
 
         $this->shouldThrow(new \InvalidArgumentException('Zone with code "EU" does not exist'))->during('defaultTaxZoneIs', ['EU']);
     }
+
+    function it_casts_zone_code_to_zone($zoneRepository, ZoneInterface $zone)
```
Comment (pjedrzejewski) [inline src/Sylius/Behat/spec/Context/Setup/ZoneContextSpec.php]: It returns zone by its code

File: src/Sylius/Behat/spec/Context/Setup/ZoneContextSpec.php [hunk 38]
```diff
@@ -79,4 +100,32 @@ function it_throws_exception_if_zone_with_given_code_does_not_exist($zoneReposit
...
+    {
+        $zoneRepository->findOneBy(['code' => 'EU'])->willReturn($zone);
+
+        $this->getZoneByCode('EU')->shouldReturn($zone);
+    }
+
+    function it_throws_exception_if_there_is_no_zone_with_given_code_to_cast($zoneRepository)
```
Comment (pjedrzejewski) [inline src/Sylius/Behat/spec/Context/Setup/ZoneContextSpec.php]: it throws exception if zone with given code does not exist

File: src/Sylius/Behat/spec/Context/Setup/ZoneContextSpec.php [hunk 39]
```diff
@@ -79,4 +100,32 @@ function it_throws_exception_if_zone_with_given_code_does_not_exist($zoneReposit
...
+    {
+        $zoneRepository->findOneBy(['code' => 'EU'])->willReturn(null);
+
+        $this->shouldThrow(new \Exception('Zone with code "EU" does not exist'))->during('getZoneByCode', ['EU']);
+    }
+
+    function it_casts_rest_of_the_world($zoneRepository, ZoneInterface $zone)
```
Comment (pjedrzejewski) [inline src/Sylius/Behat/spec/Context/Setup/ZoneContextSpec.php]: it returns the rest of the world zone

File: features/shop/apply_correct_shipping_fee_with_taxes_on_order.feature (hunk 27 above)
Comment (Zales0123) [inline features/shop/apply_correct_shipping_fee_with_taxes_on_order.feature]: :+1:

File: src/Sylius/Behat/Context/Setup/TaxContext.php (hunk 31 above)
Comment (Zales0123) [inline src/Sylius/Behat/Context/Setup/TaxContext.php]: No reason at all :smile:

File: src/Sylius/Behat/Context/Ui/CheckoutContext.php (hunk 33 above)
Comment (Zales0123) [inline src/Sylius/Behat/Context/Ui/CheckoutContext.php]: Agreed! Should be changed definitely in the nearest future ;)

File: src/Sylius/Behat/spec/Context/Setup/ChannelContextSpec.php (hunk 34 above)
Comment (Zales0123) [inline src/Sylius/Behat/spec/Context/Setup/ChannelContextSpec.php]: :+1: especially it is written this way in other specs.

File: src/Sylius/Behat/Context/Setup/TaxContext.php (hunk 31 above)
Comment (Zales0123) [inline src/Sylius/Behat/Context/Setup/TaxContext.php]: Changed my mind, it is necessary: to handle ``(the rest of the world)`` with transformer, step definition must be regex ;/

**Comment by pjedrzejewski** [comment]
Thank you Mateusz!
